aom/src/main/java/com/nedap/archie/aom/ArchetypeHRID.java

```diff
@@ -49,8 +49,8 @@
     private String buildCount;
     //TODO: XML attribute 'physical id', which is the full id
 
-    private static final Pattern namespacePattern = Pattern.compile("((?<namespace>.*)::)?");
-    private static final Pattern publisherPattern = Pattern.compile("(?<publisher>[^.-]*)");
+    private static final Pattern namespacePattern = Pattern.compile("((?<namespace>([a-zA-Z][a-zA-Z0-9-]*(?<!-))(\\.(?!-)[A-Za-z0-9-]+(?<!-))*)::)?");
```

Comment:
Just to check, In the ADL lexer grammar, this is:

```antlr
fragment NAMESPACE : LABEL ('.' LABEL)* ;
fragment LABEL : ALPHA_CHAR (NAME_CHAR|URI_PCT_ENCODED)* ;

fragment NAME_CHAR     : WORD_CHAR | '-' ;
fragment WORD_CHAR     : ALPHANUM_CHAR | '_' ;
fragment ALPHANUM_CHAR : ALPHA_CHAR | DIGIT ;
fragment ALPHA_CHAR  : [a-zA-Z] ;
fragment DIGIT     : [0-9] ;
```

For some reason, this deviates from the ADL 2 specification, which has this format:

```
fragment NAMESPACE : LABEL ('.' LABEL)+ ;
fragment LABEL : ALPHA_CHAR ( NAME_CHAR* ALPHANUM_CHAR )? ;
```
Which has one additional constraint that a label must end with an ALPHANUM_CHAR, and does not reference percent encoding.

I would say we match this with our own grammar exactly. So, then:

- What does the (?!- and (?<! lookbehind/ahead construction do, and why is it needed?
- I think we also need to support '_'?

Probably the percent encoding is not very useful, and may have been added by mistake because the namespace label was shared with URL parsing in a previous version. I'm not sure actually, probably should be dropped from the grammar and not supported here.

It looks possible to directly translate the lexer grammar into a regular expression, which would be, minus the percent encoding:
```regex
([a-zA-Z]([a-zA-Z0-9_-]*[a-zA-Z0-9])?)(\\.([a-zA-Z]([a-zA-Z0-9_-]*[a-zA-Z0-9])?))*
```

Isn't that a better solution?

Reply:
I did it like this because we got some remarks on that the namespace should conform with being a reverse domain name: https://specifications.openehr.org/releases/AM/latest/Identification.html#_formal_model & https://specifications.openehr.org/releases/AM/latest/ADL2.html#_human_readable_archetype_identifier. 

I found the following for the rules of a domain name: https://help.fasthosts.co.uk/app/answers/detail/a_id/259/~/domain-registration-rules-and-restrictions. So that was the basis I took.

According to these rules, only letters, numbers and hyphens are supported. And hyphens should not be at the beginning or end, so that's why the lookbehind/ahead construction was for.

I feel a little bit like the openEHR specifications only advises to using a reverse domain name instead of requiring it. So maybe we can be less strict, but what do you think?